dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -318,32 +371,313 @@ private List<Event> inlineAtomicEnd(FunctionCall ignored) {
+        final var error = new INonDet(constantId++, (IntegerType) result.getType(), true);
+        call.getFunction().getProgram().addConstant(error);
+        final Expression zero = expressions.makeGeneralZero(result.getType());
+        final Expression minusOne = expressions.makeValue(BigInteger.ONE.negate(), types.getArchType());
+        return List.of(
+                // Allow other threads to access the condition variable.
+                EventFactory.Pthread.newUnlock(lock.toString(), lock),
+                // Decide success
+                //TODO proper error code: ETIMEDOUT
+                EventFactory.newLocal(result, error),
+                EventFactory.newJump(expressions.makeNEQ(error, zero), label),
+                // Wait for signal or broadcast.
+                EventFactory.Atomic.newFADD(dummy, address, minusOne, Tag.C11.MO_RELAXED),
+                EventFactory.newAbortIf(expressions.makeGT(dummy, zero, true)),
+                // Wait for all waiters to be notified of a broadcast.  If signal, wait for itself.
+                // Subsequent waits cannot be notified by the same broadcast event.
+                EventFactory.Atomic.newLoad(dummy, address, Tag.C11.MO_RELAXED),
+                EventFactory.newAbortIf(expressions.makeEQ(dummy, zero)),
+                // Join branches
+                label,
+                // Re-lock.
+                EventFactory.Pthread.newLock(lock.toString(), lock));
+    }
+
+    private List<Event> inlinePthreadKeyCreate(FunctionCall call) {
+        final Register result = checkValueAndArguments(2, call);
+        final Expression key = call.getArguments().get(0);
+        final Expression destructor = call.getArguments().get(1);
+        final Program program = call.getFunction().getProgram();
+        final int threadCount = program.getThreads().size();
+        final int pointerBytes = types.getMemorySizeInBytes(types.getPointerType());
+        final MemoryObject object = program.getMemory().allocate((threadCount + 1) * pointerBytes, false);
+        final BigInteger destructorOffsetValue = BigInteger.valueOf((long) threadCount * pointerBytes);
+        final Expression destructorOffset = expressions.makeValue(destructorOffsetValue, types.getArchType());
+        //TODO call destructor at each thread's normal exit
+        return List.of(
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())),
+                EventFactory.newStore(key, object),
+                EventFactory.newStore(expressions.makeADD(object, destructorOffset), destructor));
+    }
+
+    private List<Event> inlinePthreadKeyDelete(FunctionCall call) {
+        final Register result = checkValueAndArguments(1, call);
+        //final Expression key = call.getArguments().get(0);
+        //final int threadID = call.getThread().getId();
+        return List.of(
+                EventFactory.newLocal(result, expressions.makeGeneralZero(result.getType())));
+    }
```

Comment:
The implementation could treat zero as values of uninitialized objects, instead.  Then, destructors can just check non-zero, then store zero.  This could be done for the entire pthread library.